File: plot/display_size_and_resolution.f90
!+
! Subroutine display_size_and_resolution (ix_screen, x_size, y_size, x_res, y_res)
!
! Routine to return the size and resolution of a display screen.
! Note: All output numbers will be set to zero if there is a problem obtaining the display information.
! 
! Input:
!    ix_screen -- Screen index. Generally this should be 0.
!
! Output:
!    x_size  -- real(rp): Horizontal screen size in mm.
!    y_size  -- real(rp): Vertical screen size in mm.
!    x_res   -- real(rp): Horizontal resolution in pixels / mm.
!    y_res   -- real(rp): Vertical resolution in pixels / mm.
!-


subroutine display_size_and_resolution (ix_screen, x_size, y_size, x_res, y_res)

use sim_utils, dummy => display_size_and_resolution

implicit none

interface
  subroutine display_size_and_res(ix_screen, x_size, y_size, x_res, y_res) bind (c)
    use, intrinsic :: iso_c_binding
    real(c_double) x_size, y_size, x_res, y_res
    integer(c_int), value :: ix_screen
  end subroutine
end interface

real(rp) x_size, y_size, x_res, y_res
real(rp) :: dpi
integer ix_screen, s, l, ios
character(*), parameter :: r_name = 'display_size_and_resolution'

!

call display_size_and_res(ix_screen, x_size, y_size, x_res, y_res)

call get_environment_variable('ACC_DPI_RESOLUTION', length=l, status=s)
if (s == 0) then
  block
    character(l) :: dpi_str
    call get_environment_variable('ACC_DPI_RESOLUTION', value=dpi_str)
    read (dpi_str, *, iostat = ios) dpi
    if (ios == 0) then
      x_res = dpi / 25.4_rp
      y_res = dpi / 25.4_rp
    else
      call out_io (s_error$, r_name, 'ACC_DPI_RESOLUTION ENVIRONMENT VARIABLE IS NOT A REAL NUMBER: ' // dpi_str)
    endif
  end block
end if

end subroutine

File: sim_utils/plot/display_size_and_resolution.c
//+
// void display_size_and_res(int ix_screen, double* x_size, double* y_size, double* x_res, double* y_res)
//
// Routine to return the size and resolution of a display screen.
// Note: All output numbers will be set to zero if there is a problem obtaining the display information.
// 
// Input:
//    ix_screen -- Screen index. Generally this should be 0.
//
// Output:
//    x_size  -- Horizontal screen size in mm.
//    y_size  -- Vertical screen size in mm.
//    x_res   -- Horizontal resolution in pixels / mm.
//    y_res   -- Vertical resolution in pixels / mm.
//-

// No X11 case --------------------------------------------------------------

#if defined (CESR_NOPLOT)

void display_size_and_res(int ix_screen, double* x_size, double* y_size, double* x_res, double* y_res)
{
  *x_size = 0;
  *y_size = 0;
  *x_res = 0;
  *y_res = 0;
}

// Have X11 case --------------------------------------------------------------

#else

#if defined(CESR_WINCVF)
#include <X11/Xwindows.h>
#endif

// Fix for OSX X11 issue with conda-forge build
#include <X11/Xfuncproto.h>
#ifndef _X_SENTINEL
# define _X_SENTINEL(x)
#endif
#ifndef _X_DEPRECATED
# define _X_DEPRECATED
#endif

#include <X11/Xlib.h>
#include <stdio.h>
#include <stdlib.h>

void display_size_and_res(int ix_screen, double* x_size, double* y_size, double* x_res, double* y_res)
{

  *x_size = 0;
  *y_size = 0;
  *x_res = 0;
  *y_res = 0;

  Display *dpy; 
  char *displayname = NULL;            /* server to contact */
  dpy = XOpenDisplay (displayname);
  if (!dpy) return;
  if (ix_screen >= 0 && ix_screen < ScreenCount(dpy)) {

    //  printf ("name of display:    %s\n", DisplayString (dpy));
    //  printf ("default screen number:    %d\n", DefaultScreen (dpy));
    //  printf ("number of screens:    %d\n", ScreenCount (dpy));

    *x_size = XDisplayWidthMM(dpy, ix_screen);
    *y_size = XDisplayHeightMM (dpy, ix_screen);
    *x_res = (double) XDisplayWidth(dpy, ix_screen) / XDisplayWidthMM(dpy, ix_screen);
    *y_res = (double) XDisplayHeight(dpy, ix_screen) / XDisplayHeightMM(dpy, ix_screen);
  }

  XCloseDisplay(dpy);
}

#endif
